src/breakpoints.jl

```diff
@@ -1,26 +1,103 @@
-const _breakpoints = BreakpointRef[]
+const _breakpoints = AbstractBreakpoint[]
 breakpoints() = copy(_breakpoints)
 
-Base.getindex(bp::BreakpointRef) = bp.framecode.breakpoints[bp.stmtidx]
-function Base.setindex!(bp::BreakpointRef, isactive::Bool)
-    bp.framecode.breakpoints[bp.stmtidx] = BreakpointState(isactive, bp[].condition)
+function add_to_existing_framecodes(bp::AbstractBreakpoint)
+    for framecode in values(framedict)
+        add_breakpoint_if_match!(framecode, bp)
+    end
 end
-function toggle!(bp::BreakpointRef)
-    state = bp[]
-    bp.framecode.breakpoints[bp.stmtidx] = BreakpointState(!state.isactive, state.condition)
+
+function add_breakpoint_if_match!(framecode::FrameCode, bp::AbstractBreakpoint)
+    if framecode_matches_breakpoint(framecode, bp)
+        stmtidx = bp.line === 0 ? 1 : statementnumber(framecode, bp.line)
+        breakpoint!(framecode, stmtidx, bp.condition)
```

Comment:
Do we need to check if there is already a breakpoint there? I worry that running `add_to_existing_framecodes` will reinstate `remove`d or `inactive` breakpoints.

Reply:
For full generality, each statement should be able to have arbitrary number of independent breakpoints which is doable but maybe not worth the effort.

I don't really see why overwriting that `BreakpointState` is worse than not doing anything.

Reply:
To be clear I'm simply asking what happens under circumstances like the following:
- `breakpoint(f)`
- disable that breakpoint
- define a new method for `f`
- debug either the new method or one of the old ones.

IIUC (which I may not), I am concerned that the breakpoint will be in `_breakpoints` without any record of the fact that you disabled it, so it will get added in active form. I am guessing this might be a problem for the new method but not the old ones? In which case my "check if there is already a breakpoint there" is a red herring---what I am really after is whether we're recording state everywhere we need it. Seems like you might need an `active` field in `BreakpointSignature` and `BreakpointFileLocations`?

Reply:
> Seems like you might need an active field in BreakpointSignature and BreakpointFileLocations?

Yes, this sounds correct, gotcha. Will make sure doing those steps does something sensible.

Reply:
Ok, should be fixed, with a test added.